Liquibase loadData fails to insert UUID in mysql

### Search first

- [X] I searched and no similar issues were found

### Description

When using loadData to load a UUID into a column in MySQL setting the type of the column in loadData causes the following failure when inserting the data:

`Caused by: liquibase.exception.DatabaseException: java.sql.SQLException: Incorrect string value: '\xAC\xED\x00\x05sr...' for column 'policy_id' at row 1`

If you change the type of the column in loadData to `STRING` this data is able to load, but that can cause issues for other dbms that need different special handling for uuids. In particular, the documentation here https://docs.liquibase.com/workflows/liquibase-community/working-with-uuids.html implies that using uuid as the conversion type is the correct one.

### Steps To Reproduce

I created a table with initial data on a MySQL database using the following files:

```
databaseChangeLog:
- changeSet:
    author: asayman
    changes:
    - createTable:
        columns:
        - column:
            constraints:
              nullable: false
              primaryKey: true
              unique: false
            name: policy_id
            type: UUID
        tableName: ProductPolicy
    - modifySql:
        dbms: mysql
        replace:
          replace: NVARCHAR(100)
          with: VARCHAR(100) CHARACTER SET UTF8MB4
    - loadData:
        columns:
        - column:
            header: policy_id
            name: policy_id
            type: uuid
        encoding: UTF-8
        file: ../changelog-initial-data/001-product-policy.csv
        quotechar: ''''
        relativeToChangelogFile: true
        separator: ','
        tableName: ProductPolicy
    id: 000-table-productpolicy
```

```
policy_id
1872003c-2b7c-4678-994a-0ff3c24f4797
e231b4aa-3d37-4736-9459-1d6f1e469e33
825bf43d-45fd-4a49-8820-1164bde70830
0fef3af7-efbd-4ef1-91b6-10ea32e02c9f
e1c07c47-59ec-4f14-b6a2-a1879c764d97
f0d1243b-0d3b-4b14-8336-9cd1c4de3176
651e2feb-46b9-4983-b463-dad8e2ff6816
5fdb4863-0872-4b69-9213-9938cd9c138f
```

which results in the following error:

```
####################################################
##   _     _             _ _                      ##
##  | |   (_)           (_) |                     ##
##  | |    _  __ _ _   _ _| |__   __ _ ___  ___   ##
##  | |   | |/ _` | | | | | '_ \ / _` / __|/ _ \  ##
##  | |___| | (_| | |_| | | |_) | (_| \__ \  __/  ##
##  \_____/_|\__, |\__,_|_|_.__/ \__,_|___/\___|  ##
##              | |                               ##
##              |_|                               ##
##                                                ## 
##  Get documentation at docs.liquibase.com       ##
##  Get certified courses at learn.liquibase.com  ## 
##                                                ##
####################################################
Starting Liquibase at 14:46:31 (version 4.24.0 #14062 built at 2023-09-28 12:18+0000)
Liquibase Version: 4.24.0
Liquibase Open Source 4.24.0 by Liquibase
Running Changeset: changelog-tables/050-052-table-productpolicy.yaml::050-052-000-table-productpolicy::asayman

Unexpected error running Liquibase: Migration failed for changeset changelog-tables/050-052-table-productpolicy.yaml::050-052-000-table-productpolicy::asayman:
     Reason: liquibase.exception.DatabaseException: java.sql.SQLException: Incorrect string value: '\xAC\xED\x00\x05sr...' for column 'policy_id' at row 1
```

If you change the above yaml file to the following, there is no error:

```
databaseChangeLog:
- changeSet:
    author: asayman
    changes:
    - createTable:
        columns:
        - column:
            constraints:
              nullable: false
              primaryKey: true
              unique: false
            name: policy_id
            type: UUID
        tableName: ProductPolicy
    - modifySql:
        dbms: mysql
        replace:
          replace: NVARCHAR(100)
          with: VARCHAR(100) CHARACTER SET UTF8MB4
    - loadData:
        columns:
        - column:
            header: policy_id
            name: policy_id
            type: STRING
        encoding: UTF-8
        file: ../changelog-initial-data/001-product-policy.csv
        quotechar: ''''
        relativeToChangelogFile: true
        separator: ','
        tableName: ProductPolicy
    id: 050-052-000-table-productpolicy
```


### Expected/Desired Behavior

There are no errors and the rows correctly have the UUID specified in the csv file.

### Liquibase Version

4.24.0

### Database Vendor & Version

MySQL Server 8.1.0-1.el8

### Liquibase Integration

_No response_

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

Ubuntu 22.04.3 LTS

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

--- a/liquibase-integration-tests/src/test/java/liquibase/dbtest/mysql/MySQLIntegrationTest.java
+++ b/liquibase-integration-tests/src/test/java/liquibase/dbtest/mysql/MySQLIntegrationTest.java
@@ -1,11 +1,14 @@
 package liquibase.dbtest.mysql;
 
 import liquibase.CatalogAndSchema;
+import liquibase.Liquibase;
 import liquibase.Scope;
 import liquibase.database.DatabaseFactory;
+import liquibase.datatype.DataTypeFactory;
 import liquibase.dbtest.AbstractIntegrationTest;
 import liquibase.exception.DatabaseException;
 import liquibase.executor.ExecutorService;
+import liquibase.resource.Resource;
 import liquibase.snapshot.DatabaseSnapshot;
 import liquibase.snapshot.SnapshotControl;
 import liquibase.snapshot.SnapshotGeneratorFactory;
@@ -13,12 +16,20 @@ import liquibase.statement.core.RawParameterizedSqlStatement;
 import liquibase.structure.core.Column;
 import liquibase.structure.core.Schema;
 import liquibase.structure.core.Table;
+import liquibase.test.JUnitResourceAccessor;
 import org.junit.Test;
 
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.nio.ByteBuffer;
 import java.sql.SQLSyntaxErrorException;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assume.assumeNotNull;
 import static org.junit.Assume.assumeTrue;
 
 /**
@@ -43,8 +54,13 @@ import static org.junit.Assume.assumeTrue;
  */
 public class MySQLIntegrationTest extends AbstractIntegrationTest {
 
+    private final String uuidTableChangelog;
+    private final String uuidLoadDataCSV;
+
     public MySQLIntegrationTest() throws Exception {
         super("mysql", DatabaseFactory.getInstance().getDatabase("mysql"));
+        uuidTableChangelog = "changelogs/mysql/complete/uuid.changelog.xml";
+        uuidLoadDataCSV = "changelogs/mysql/complete/datafiles/uuid-data.csv";
     }
 
     @Test
@@ -85,6 +101,40 @@ public class MySQLIntegrationTest extends AbstractIntegrationTest {
         Object defaultValue = createdColumn.getDefaultValue();
         assertNotNull(defaultValue);
         assertEquals("0000-00-00 00:00:00", defaultValue);
+    }
+
+    @Test
+    public void testUUIDCreateLoadDataChangelog() throws Exception {
+        assumeNotNull(this.getDatabase());
+        clearDatabase();
+        // update database to create table
+        Liquibase liquibase = createLiquibase(uuidTableChangelog);
+        liquibase.update();
+        DatabaseSnapshot snapshot = SnapshotGeneratorFactory.getInstance().createSnapshot(CatalogAndSchema.DEFAULT, this.getDatabase(), new SnapshotControl(getDatabase()));
+
+        Table uuidTable = snapshot.get(Table.class).stream().filter(t -> t.getName().equals("uuidtest")).findFirst().orElse(null);
+
+        // assert created table and created 1 column
+        assertNotNull(uuidTable);
+        assertEquals(1, uuidTable.getColumns().size());
+
+        // assert created column is named uuid, and is of type BINARY(16) for MySQL UUID.
+        Column uuidColumn = uuidTable.getColumns().get(0);
+        assertEquals("uuid", uuidColumn.getName());
+        assertEquals(
+                DataTypeFactory.getInstance().fromDescription("char(36)", getDatabase()),
+                DataTypeFactory.getInstance().fromDescription(uuidColumn.getType().toString(), getDatabase()));
+
+        // assert that the UUID data was loaded correctly
+        List<Map<String, ?>> queryResult = Scope.getCurrentScope().getSingleton(ExecutorService.class).getExecutor("jdbc", getDatabase())
+                .queryForList(new RawParameterizedSqlStatement(String.format("select * from %s", uuidTable.getName())));
+
+        Resource fileOpener = new JUnitResourceAccessor().get(uuidLoadDataCSV);
+        BufferedReader br = new BufferedReader(new FileReader(fileOpener.getUri().getPath()));
+        br.readLine();
+        String expected = br.readLine();
+        String actual = (String) queryResult.get(0).get("uuid");
+        assertEquals(expected, actual);
     }
 
 }

--- a/liquibase-integration-tests/src/test/java/liquibase/dbtest/mysql/MySQLIntegrationTest.java
+++ b/liquibase-integration-tests/src/test/java/liquibase/dbtest/mysql/MySQLIntegrationTest.java
@@ -21,11 +21,9 @@ import org.junit.Test;
 
 import java.io.BufferedReader;
 import java.io.FileReader;
-import java.nio.ByteBuffer;
 import java.sql.SQLSyntaxErrorException;
 import java.util.List;
 import java.util.Map;
-import java.util.UUID;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
@@ -118,7 +116,7 @@ public class MySQLIntegrationTest extends AbstractIntegrationTest {
         assertNotNull(uuidTable);
         assertEquals(1, uuidTable.getColumns().size());
 
-        // assert created column is named uuid, and is of type BINARY(16) for MySQL UUID.
+        // assert created column is named uuid, and is of type char(36) for MySQL UUID.
         Column uuidColumn = uuidTable.getColumns().get(0);
         assertEquals("uuid", uuidColumn.getName());
         assertEquals(

--- a/liquibase-integration-tests/src/test/java/liquibase/dbtest/mysql/MySQLIntegrationTest.java
+++ b/liquibase-integration-tests/src/test/java/liquibase/dbtest/mysql/MySQLIntegrationTest.java
@@ -129,12 +129,15 @@ public class MySQLIntegrationTest extends AbstractIntegrationTest {
         List<Map<String, ?>> queryResult = Scope.getCurrentScope().getSingleton(ExecutorService.class).getExecutor("jdbc", getDatabase())
                 .queryForList(new RawParameterizedSqlStatement(String.format("select * from %s", uuidTable.getName())));
 
-        Resource fileOpener = new JUnitResourceAccessor().get(uuidLoadDataCSV);
-        BufferedReader br = new BufferedReader(new FileReader(fileOpener.getUri().getPath()));
-        br.readLine();
-        String expected = br.readLine();
-        String actual = (String) queryResult.get(0).get("uuid");
-        assertEquals(expected, actual);
+        try (JUnitResourceAccessor jra = new JUnitResourceAccessor()) {
+            Resource fileOpener = jra.get(uuidLoadDataCSV);
+            try (BufferedReader br = new BufferedReader(new FileReader(fileOpener.getUri().getPath()))) {
+                br.readLine();
+                String expected = br.readLine();
+                String actual = (String) queryResult.get(0).get("uuid");
+                assertEquals(expected, actual);
+            }
+        }
     }
 
 }

--- a/liquibase-integration-tests/src/test/resources/changelogs/mysql/complete/datafiles/uuid-data.csv
+++ b/liquibase-integration-tests/src/test/resources/changelogs/mysql/complete/datafiles/uuid-data.csv
@@ -1,0 +1,2 @@
+uuid
+339a5f0c-d77f-45ba-b801-3325cf8f32ee

--- a/liquibase-integration-tests/src/test/resources/changelogs/mysql/complete/uuid.changelog.xml
+++ b/liquibase-integration-tests/src/test/resources/changelogs/mysql/complete/uuid.changelog.xml
@@ -1,0 +1,17 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd">
+    <changeSet id="uuid-1" author="qsevers">
+        <createTable tableName="uuidtest">
+            <column name="uuid" type="uuid">
+                <constraints primaryKey="true"/>
+            </column>
+        </createTable>
+    </changeSet>
+
+    <changeSet id="uuid-2" author="qsevers">
+        <loadData tableName="uuidtest" file="changelogs/mysql/complete/datafiles/uuid-data.csv"/>
+    </changeSet>
+
+</databaseChangeLog>

--- a/liquibase-standard/src/main/java/liquibase/statement/ExecutablePreparedStatementBase.java
+++ b/liquibase-standard/src/main/java/liquibase/statement/ExecutablePreparedStatementBase.java
@@ -6,6 +6,8 @@ import liquibase.change.core.LoadDataChange;
 import liquibase.changelog.ChangeSet;
 import liquibase.database.Database;
 import liquibase.database.PreparedStatementFactory;
+import liquibase.database.core.MariaDBDatabase;
+import liquibase.database.core.MySQLDatabase;
 import liquibase.database.core.PostgresDatabase;
 import liquibase.database.core.SQLiteDatabase;
 import liquibase.datatype.DataTypeFactory;
@@ -32,6 +34,7 @@ import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.math.BigDecimal;
 import java.math.BigInteger;
+import java.nio.ByteBuffer;
 import java.sql.*;
 import java.util.*;
 import java.util.logging.Level;
@@ -241,7 +244,11 @@ public abstract class ExecutablePreparedStatementBase implements ExecutablePrepa
         if (col.getValue() != null) {
             LOG.fine("value is string/UUID/blob = " + col.getValue());
             if (col.getType() != null && col.getType().equalsIgnoreCase(LoadDataChange.LOAD_DATA_TYPE.UUID.name())) {
-                stmt.setObject(i, UUID.fromString(col.getValue()));
+                if (database instanceof MySQLDatabase && !(database instanceof MariaDBDatabase))  {
+                    stmt.setString(i, col.getValue());
+                } else {
+                    stmt.setObject(i, UUID.fromString(col.getValue()));
+                }
             } else if (col.getType() != null && col.getType().equalsIgnoreCase(LoadDataChange.LOAD_DATA_TYPE.OTHER.name())) {
                 stmt.setObject(i, col.getValue(), Types.OTHER);
             } else if (LoadDataChange.LOAD_DATA_TYPE.BLOB.name().equalsIgnoreCase(col.getType())) {

